proxy/src/control/discovery.rs

```diff
@@ -506,7 +602,107 @@ where T: HttpService<RequestBody = BoxBody, ResponseBody = RecvBody>,
     }
 }
 
-// ===== impl RxError =====
+// ===== impl DstLabels ====
+
+impl DstLabels {
+
+    #[inline]
+    fn new<I, S>(labels: I) -> Option<Self>
+    where
+        I: IntoIterator<Item=(S, S)>,
+        S: fmt::Display,
+    {
+        let mut labels = labels.into_iter();
+
+        if let Some((k, v)) = labels.next() {
+            // format the first label pair without a leading comma, since we
+            // don't know where it is in the output labels at this point.
+            let mut s = format!("dst_{}=\"{}\"", k, v);
+
+            // format subsequent label pairs with leading commas, since
+            // we know that we already formatted the first label pair.
+            for (k, v) in labels {
+                write!(s, ",dst_{}=\"{}\"", k, v)
+                    .expect("writing to string should not fail");
+            }
+
+            Some(DstLabels(Arc::from(s)))
+        } else {
+            // the iterator is empty; return none
+            None
+        }
+
+    }
+
+}
+
+impl fmt::Display for DstLabels {
+
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+
+}
+
+// ===== impl Labeled =====
+
+impl<T> Labeled<T> {
+
+    /// Wrap `inner` with a `Watch` on dyanmically updated labels.
+    pub fn new(inner: T,
+               watch: futures_watch::Watch<Option<DstLabels>>)
+               -> Self
+    {
+        Self {
+            metric_labels: Some(watch),
+            inner,
+        }
+    }
+
+    /// Wrap `inner` with no `metric_labels`.
+    pub fn none(inner: T) -> Self {
+        Self { metric_labels: None, inner }
+    }
+}
+
+impl<T, A> Service for Labeled<T>
+where
+    T: Service<Request=http::Request<A>>,
+{
+    type Request = T::Request;
+    type Response= T::Response;
+    type Error = T::Error;
+    type Future = T::Future;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, req: Self::Request) -> Self::Future {
+        let mut req = req;
+        if let Some(labels) = self.metric_labels.as_ref()
+            .and_then(|labels| (*labels.borrow()).as_ref().cloned())
+        {
+            req.extensions_mut().insert(labels);
+        }
+        self.inner.call(req)
```

Comment:
it would be great to write a unit test for this middleware, but lets do that in a followup.